File: trax_app/lib/models/trail.dart
class Trail {
  final String? id;
  final String name;
  final String? location;
  final String difficulty; // easy, medium, hard, extreme
  final bool isPublic;
  final String? description;
  final String? imageUrl;
  final double? distance; // km
  final double? elevation; // meters (cumulative gain)
  final double? elevationDiff; // meters (max altitude - min altitude)
  final String? type; // free_ride, lap
  final int? creatorId;
  final String? creatorName;
  final String? creatorAvatarUrl;
  final double? startLatitude;
  final double? startLongitude;
  final double? endLatitude;
  final double? endLongitude;
  final DateTime? createdAt;

  Trail({
    this.id,
    required this.name,
    this.location,
    this.difficulty = 'medium',
    this.isPublic = true,
    this.description,
    this.imageUrl,
    this.distance,
    this.elevation,
    this.elevationDiff,
    this.type,
    this.creatorId,
    this.creatorName,
    this.creatorAvatarUrl,
    this.startLatitude,
    this.startLongitude,
    this.endLatitude,
    this.endLongitude,
    this.createdAt,
  });

  factory Trail.fromJson(Map<String, dynamic> json) {
    return Trail(
      id: json['id']?.toString(),
      name: json['name'] ?? '',
      location: json['location'],
      difficulty: json['difficulty'] ?? 'medium',
      isPublic: json['public'] ?? true,
      description: json['description'],
      imageUrl: json['imageUrl'],
      distance: (json['distance'] as num?)?.toDouble(),
      elevation: (json['elevation'] as num?)?.toDouble(),
      elevationDiff: (json['elevationDiff'] as num?)?.toDouble(),
      type: json['type'],
      creatorId: (json['creatorId'] as num?)?.toInt(),
      creatorName: json['creatorName'],
      creatorAvatarUrl: json['creatorAvatarUrl'],
      startLatitude: (json['startLatitude'] as num?)?.toDouble(),
      startLongitude: (json['startLongitude'] as num?)?.toDouble(),
      endLatitude: (json['endLatitude'] as num?)?.toDouble(),
      endLongitude: (json['endLongitude'] as num?)?.toDouble(),
      createdAt: json['createdAt'] != null ? DateTime.tryParse(json['createdAt']) : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'location': location,
        'difficulty': difficulty,
        'isPublic': isPublic,
        'description': description,
        'imageUrl': imageUrl,
        'distance': distance,
        'elevation': elevation,
        'elevationDiff': elevationDiff,
        'type': type,
        'creatorId': creatorId,
        'creatorName': creatorName,
        'creatorAvatarUrl': creatorAvatarUrl,
        'startLatitude': startLatitude,
        'startLongitude': startLongitude,
        'endLatitude': endLatitude,
        'endLongitude': endLongitude,
        'createdAt': createdAt?.toIso8601String(),
      };
}

File: backend/src/main/java/com/trax/model/Trail.java
package com.trax.model;

import jakarta.persistence.*;
import java.time.LocalDateTime;

@Entity
@Table(name = "trails")
public class Trail {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    private String location;
    private String difficulty; // EASY, MEDIUM, HARD, EXTREME
    private String description;
    private String imageUrl;
    private Double distance;
    private Double elevation;
    private Double elevationDiff;
    private String type; // free_ride, lap
    private Long creatorId;
    private Double startLatitude;
    private Double startLongitude;
    private Double endLatitude;
    private Double endLongitude;
    private LocalDateTime createdAt;
    @Column(nullable = false, columnDefinition = "BOOLEAN DEFAULT FALSE")
    private boolean deleted = false;
    @Column(nullable = false, columnDefinition = "BOOLEAN DEFAULT TRUE")
    private boolean isPublic = true;

    @PrePersist
    protected void onCreate() {
        if (createdAt == null) createdAt = LocalDateTime.now();
    }

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    public String getLocation() { return location; }
    public void setLocation(String location) { this.location = location; }
    public String getDifficulty() { return difficulty; }
    public void setDifficulty(String difficulty) { this.difficulty = difficulty; }
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    public String getImageUrl() { return imageUrl; }
    public void setImageUrl(String imageUrl) { this.imageUrl = imageUrl; }
    public Double getDistance() { return distance; }
    public void setDistance(Double distance) { this.distance = distance; }
    public Double getElevation() { return elevation; }
    public void setElevation(Double elevation) { this.elevation = elevation; }
    public Double getElevationDiff() { return elevationDiff; }
    public void setElevationDiff(Double elevationDiff) { this.elevationDiff = elevationDiff; }
    public String getType() { return type; }
    public void setType(String type) { this.type = type; }
    public Long getCreatorId() { return creatorId; }
    public void setCreatorId(Long creatorId) { this.creatorId = creatorId; }
    public Double getStartLatitude() { return startLatitude; }
    public void setStartLatitude(Double startLatitude) { this.startLatitude = startLatitude; }
    public Double getStartLongitude() { return startLongitude; }
    public void setStartLongitude(Double startLongitude) { this.startLongitude = startLongitude; }
    public Double getEndLatitude() { return endLatitude; }
    public void setEndLatitude(Double endLatitude) { this.endLatitude = endLatitude; }
    public Double getEndLongitude() { return endLongitude; }
    public void setEndLongitude(Double endLongitude) { this.endLongitude = endLongitude; }
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    public boolean isPublic() { return isPublic; }
    public void setPublic(boolean isPublic) { this.isPublic = isPublic; }
    public boolean isDeleted() { return deleted; }
    public void setDeleted(boolean deleted) { this.deleted = deleted; }
}
